File: code/data_structures/hashtable/HashTableDoubleHashing.py

from HashTableOpenAddressingBase import HashTableOpenAddressingBase

from abc import ABC, abstractmethod

from typing import TypeVar

K = TypeVar('K', bound='SecondaryHash')
V = TypeVar('V')


class SecondaryHash(ABC):
    @abstractmethod
    def hash_code2(self) -> int:
        pass

class HashTableDoubleHashing(HashTableOpenAddressingBase[K, V]):

    def __init__(self, capacity=7, load_factor=0.65):
        super().__init__(capacity, load_factor)
        self.hash = 0

    def setup_probing(self, key):
        self.hash = self.normalize_index(key.hash_code2())

        if self.hash == 0:
            self.hash = 1

    def probe(self, x):
        return x * self.hash

    def adjust_capacity(self):
        while not self.is_probable_prime(self.capacity):
            self.capacity += 1

    @staticmethod
    def is_probable_prime(n, k=20):
        return pow(2, n - 1, n) == 1 if n >= 3 else n == 2

File: code/data_structures/hashtable/HashTableDoubleHashing.cpp


#include <iostream>
#include <vector>
#include <stdexcept>
#include <cmath>
#include "HashTableOpenAddressingBase.cpp"

template<typename K, typename V>
class HashTableDoubleHashing : public HashTableOpenAddressingBase<K, V> {
private:
    int hash;

public:
    HashTableDoubleHashing(int capacity = 7, double loadFactor = 0.65) : HashTableOpenAddressingBase<K, V>(capacity, loadFactor), hash(0) {}

    void setupProbing(const K& key) override {
        hash = this->normalizeIndex(key.hash_code2());
        if (hash == 0)
            hash = 1;
    }

    int probe(int x) override {
        return x * hash;
    }

    void adjustCapacity() override {
        while (!isProbablePrime(this->capacity))
            this->capacity++;
    }

    static bool isProbablePrime(int n, int k = 20) {
        return (n >= 3) ? (std::pow(2, n - 1) % n == 1) : (n == 2);
    }
};
